screens/locationData.js

```diff
@@ -0,0 +1,179 @@
+    padding: 10,
+    marginHorizontal: 5,
+    marginVertical: 10,
+  },
+  first: {
+    borderTopWidth: 1,
+  },
+  section: {
+    flexDirection: 'row',
+    paddingVertical: 15,
+    paddingHorizontal: 10,
+    backgroundColor: colours.greys.lightest,
+    borderBottomWidth: 1,
+    borderColor: colours.black,
+    alignItems: 'center',
+  },
+  sectionText: {
+    fontSize: fonts.large,
+  },
+});
+
+class LocationData extends PureComponent {
+  state = {
+    failedOpen: false,
+    numberUploading: 0,
+  }
+
+  onFailedPressed = () => {
+    this.setState(p => ({ failedOpen: !p.failedOpen }));
+  }
+
+  uploadLocations = () => {
+    this.setState({ numberUploading: this.props.stats.offline + this.props.stats.failed });
+    this.props.showPushDialog();
+  }
+
+  render() {
+    const {
+      failedLocations,
+      stats,
+      uploading,
+    } = this.props;
+    const {
+      failedOpen,
+      numberUploading,
+    } = this.state;
+
+    const iconName = failedOpen ? 'chevron-small-down' : 'chevron-small-right';
+    const disabled = stats.offline === 0 && stats.failed === 0;
+
+    return (
+      <View style={styles.container}>
+        <View>
+          <TouchableOpacity style={styles.button} disabled={disabled} onPress={this.uploadLocations}>
+            <Text style={styles.sectionText}>{I18n.t('settings/push_locations')}</Text>
+          </TouchableOpacity>
+        </View>
+        <View style={[styles.section, styles.first]}>
+          { uploading && <Text style={styles.sectionText}>{I18n.t('locations/uploading')}{`: ${numberUploading - stats.offline}/${numberUploading}`}</Text> }
+          { !uploading && <Text style={styles.sectionText}>{I18n.t('locations/offline')}{`: ${stats.offline + stats.failed}`}</Text> }
+        </View>
+        <View style={styles.section}>
+          <Text style={styles.sectionText}>{I18n.t('locations/successful_upload')}{`: ${stats.uploaded}`}</Text>
+        </View>
+        <View style={styles.section}>
+          <Icon size="medium" family="entypo" name={iconName} colour={colours.black} />
+          <TouchableOpacity onPress={this.onFailedPressed}>
+            <Text style={styles.sectionText}>{I18n.t('locations/Failed_upload')}{`: ${stats.failed}`}</Text>
+          </TouchableOpacity>
+        </View>
+        {stats.failed > 0 &&
+          <CollapsibleList
+            data={failedLocations}
+            open={failedOpen}
+          />
+        }
+      </View>
+    );
+  }
+}
+
+LocationData.propTypes = {
+  failedLocations: PropTypes.array,
+  showPushDialog: PropTypes.func.isRequired,
+  stats: PropTypes.shape({
+    pending: PropTypes.number.isRequired,
+    offline: PropTypes.number.isRequired,
+    failed: PropTypes.number.isRequired,
+    uploaded: PropTypes.number.isRequired,
+  }).isRequired,
+  uploading: PropTypes.bool.isRequired,
+};
+
+LocationData.defaultProps = {
+  failedLocations: [],
+};
+
+const mapStateToProps = (state) => {
```

Comment:
You might want to make a selector for this. It seems a bit fancy.

Reply:
Having a hard time doing this cleanly. Without having separate selectors for each of the UploadStatus counters in the stats object... 😕